torch/utils/_pytree.py

```diff
@@ -1181,3 +1300,91 @@ def arg_tree_leaves(*args: PyTree, **kwargs: PyTree) -> List[Any]:
     for a in kwargs.values():
         _tree_leaves_helper(a, leaves)
     return leaves
+
+
+def tree_flatten_with_path(tree: PyTree) -> Tuple[List[Tuple[KeyPath, Any]], TreeSpec]:
+    """Flattens a pytree like :func:`tree_flatten`, but also returns each leaf's key path.
+
+    Args:
+        tree: a pytree to flatten. If it contains a custom type, that type must be
+            registered with an appropriate `tree_flatten_with_path_fn` when registered
+            with :func:`register_pytree_node`.
+    Returns:
+        A tuple where the first element is a list of (key path, leaf) pairs, and the
+        second element is a :class:`TreeSpec` representing the structure of the flattened
+        tree.
+    """
+    _, treespec = tree_flatten(tree)
+    return list(_generate_key_paths((), tree)), treespec
+
+
+def tree_leaves_with_path(tree: PyTree) -> List[Tuple[KeyPath, Any]]:
+    """Gets the leaves of a pytree like ``tree_leaves`` and returns each leaf's key path.
+
+    Args:
+        tree: a pytree. If it contains a custom type, that type must be
+            registered with an appropriate `tree_flatten_with_path_fn` when registered
+            with :func:`register_pytree_node`.
+    Returns:
+        A list of (key path, leaf) pairs.
+    """
+    return list(_generate_key_paths((), tree))
+
+
+def _generate_key_paths(
+    key_path: KeyPath,
+    tree: Any,
+) -> Iterable[Tuple[KeyPath, Any]]:
```

Comment:
Need to support extra custom leaf predicate function `is_leaf`. See also:

- #116419

```suggestion
def _generate_key_paths(
    key_path: KeyPath,
    tree: Any,
    is_leaf: Optional[Callable[[PyTree], bool]] = None
) -> Iterable[Tuple[KeyPath, Any]]:
```

Reply:
awesome, didn't realize we added `is_leaf` to pytree!